core/src/main/java/org/apache/iceberg/MergingSnapshotProducer.java

```diff
@@ -202,6 +207,58 @@ private ManifestFile copyManifest(ManifestFile manifest) {
         current.formatVersion(), toCopy, current.specsById(), newManifestPath, snapshotId(), appendedManifestsSummary);
   }
 
+  /**
+   * Validates that no files matching a filter have been added to the table since a starting snapshot.
+   *
+   * @param base table metadata to validate
+   * @param startingSnapshotId id of the snapshot current at the start of the operation
+   * @param conflictDetectionFilter an expression used to find new conflicting data files
+   * @param caseSensitive whether expression evaluation should be case sensitive
+   */
+  protected void validateAddedDataFiles(TableMetadata base, Long startingSnapshotId,
+                                        Expression conflictDetectionFilter, boolean caseSensitive) {
+    List<ManifestFile> manifests = Lists.newArrayList();
+    Set<Long> newSnapshots = Sets.newHashSet();
+
+    Long currentSnapshotId = base.currentSnapshot().snapshotId();
+    while (currentSnapshotId != null && !currentSnapshotId.equals(startingSnapshotId)) {
+      Snapshot currentSnapshot = ops.current().snapshot(currentSnapshotId);
+
+      ValidationException.check(currentSnapshot != null,
+          "Cannot determine history between starting snapshot %s and current %s",
+          startingSnapshotId, currentSnapshotId);
+
+      newSnapshots.add(currentSnapshotId);
+      for (ManifestFile manifest : currentSnapshot.dataManifests()) {
+        if (manifest.snapshotId() == (long) currentSnapshotId) {
+          manifests.add(manifest);
+        }
+      }
+
+      currentSnapshotId = currentSnapshot.parentId();
+    }
+
+    ManifestGroup conflictGroup = new ManifestGroup(ops.io(), manifests, ImmutableList.of())
+        .caseSensitive(caseSensitive)
+        .filterManifestEntries(entry -> newSnapshots.contains(entry.snapshotId()))
+        .filterData(conflictDetectionFilter)
+        .specsById(base.specsById())
+        .ignoreDeleted()
+        .ignoreExisting();
```

Comment:
Do we expect the number of manifests to be small enough so that we won't have to parallelize this?

Reply:
It probably should be small if we ignore `rewrite` snapshots that may include metadata rewrites.

Reply:
You're right, we can ignore rewrite snapshots. That would be helpful.

Reply:
This was fairly easy, but I had to change a test: https://github.com/apache/iceberg/blob/master/core/src/test/java/org/apache/iceberg/TestOverwriteWithValidation.java#L404-L406

Reply:
Yes, the number of changes should be small, especially if we ignore rewrites.